### PR TITLE
Migrate dist from TH to ATen(CPU, CUDA)

### PR DESCRIPTION
**Pull request by rivergold**
[#24691](https://github.com/pytorch/pytorch/issues/24691)
[#24551](https://github.com/pytorch/pytorch/issues/24551)

Benchmark:

**Speed**
```python
import time, sys
import torch
import math

inf = math.inf

torch.manual_seed(0)
devices = ["cpu", "cuda"]
ps = [0, 1, 2, 3, 4, inf, -inf]

# Warm up
for device in devices:
    for n in [1, 10, 100, 1000]:
        x = torch.randn(100, n, requires_grad=False, device=device)
        y = torch.randn(100, n, requires_grad=False, device=device)
        for i in range(1000):
            for p in ps:
                dist_xy = torch.dist(x, y, p)

for device in devices:
    print('On {}'.format(device))
    for n in [1, 10, 100, 1000]:
        total_time = 0
        x = torch.randn(100, n, requires_grad=False, device=device)
        y = torch.randn(100, n, requires_grad=False, device=device)
        for i in range(10000):
            for p in ps:
                t1 = time.time()
                dist_xy = torch.dist(x, y, p)
                t2 = time.time()
                total_time += (t2 - t1)
        average_time = total_time / 10000 / len(ps) * 1000
        print("input size(100, %d) average time is %.8f (ms)." % (n, average_time))
```

Output
Before:
```shel
On cpu
input size(100, 1) average time is 0.0079491 (ms).
input size(100, 10) average time is 0.0364167 (ms).
input size(100, 100) average time is 0.3120752 (ms).
input size(100, 1000) average time is 3.0605820 (ms).
On cuda
input size(100, 1) average time is 0.04745627 (ms).
input size(100, 10) average time is 0.04919453 (ms).
input size(100, 100) average time is 0.06601572 (ms).
input size(100, 1000) average time is 0.07849015 (ms).
```

After:
```shell
On cpu
input size(100, 1) average time is 0.0099936 (ms).
input size(100, 10) average time is 0.0340414 (ms).
input size(100, 100) average time is 0.2793379 (ms).
input size(100, 1000) average time is 0.7858076 (ms).
On cuda
input size(100, 1) average time is 0.04410237 (ms).
input size(100, 10) average time is 0.03326339 (ms).
input size(100, 100) average time is 0.03314828 (ms).
input size(100, 1000) average time is 0.03990038 (ms).
```

**Precision**

```python
for device in devices:
    torch.manual_seed(0)
    print('On {}'.format(device))
    for n in [1, 10, 100, 1000]:
        x = torch.randn(100, n, requires_grad=False).to(device)
        y = torch.randn(100, n, requires_grad=False).to(device)
        for p in ps:
            dist_xy_float = torch.dist(x, y, p)
            dist_xy_double = torch.dist(x.double(), y.double(), p)
            difference = torch.abs(dist_xy_double - dist_xy_float)
            print('input size (100, {}), p: {}, float: {}, double: {}, difference: {}'.format(n, p, dist_xy_float, dist_xy_double, difference))
```

Part of [output](https://gist.github.com/rivergold/dd95014dc7f163b22f72699d1134cdd2) 
Before:
```shell
On cpu
input size (100, 100), p: 0, float: 10000.0, double: 10000.0, difference: 0.0
input size (100, 100), p: 1, float: 11474.1806640625, double: 11474.185433543797, difference: 0.00476948129653465
input size (100, 100), p: 2, float: 143.50729370117188, double: 143.5073391487937, difference: 4.5447621829453055e-05
input size (100, 100), p: 3, float: 36.045475006103516, double: 36.04550275212738, difference: 2.774602386779179e-05
input size (100, 100), p: 4, float: 18.796083450317383, double: 18.79609807865317, difference: 1.4628335787136848e-05
input size (100, 100), p: inf, float: 5.540258407592773, double: 5.5402586460113525, difference: 2.384185791015625e-07
input size (100, 100), p: -inf, float: 3.4868717193603516e-06, double: 3.4868717193603516e-06, difference: 0.0
On cuda
input size (100, 100), p: 0, float: 10000.0, double: 10000.0, difference: 0.0
input size (100, 100), p: 1, float: 11474.1865234375, double: 11474.185433543797, difference: 0.00108989370346535
input size (100, 100), p: 2, float: 143.50733947753906, double: 143.5073391487933, difference: 3.2874575595087663e-07
input size (100, 100), p: 3, float: 36.04550552368164, double: 36.045502752127405, difference: 2.7715542358919265e-06
input size (100, 100), p: 4, float: 18.796098709106445, double: 18.796098078653177, difference: 6.304532682577246e-07
input size (100, 100), p: inf, float: 5.540258407592773, double: 5.5402586460113525, difference: 2.384185791015625e-07
input size (100, 100), p: -inf, float: 3.4868717193603516e-06, double: 3.4868717193603516e-06, difference: 0.0
```

After
```shell
On cpu
input size (100, 100), p: 0, float: 10000.0, double: 10000.0, difference: 0.0
input size (100, 100), p: 1, float: 11474.1806640625, double: 11474.185433543797, difference: 0.00476948129653465
input size (100, 100), p: 2, float: 143.50729370117188, double: 143.5073391487937, difference: 4.5447621829453055e-05
input size (100, 100), p: 3, float: 36.045475006103516, double: 36.04550275212738, difference: 2.774602386779179e-05
input size (100, 100), p: 4, float: 18.796083450317383, double: 18.79609807865317, difference: 1.4628335787136848e-05
input size (100, 100), p: inf, float: 5.540258407592773, double: 5.5402586460113525, difference: 2.384185791015625e-07
input size (100, 100), p: -inf, float: 3.4868717193603516e-06, double: 3.4868717193603516e-06, difference: 0.0
On cuda
input size (100, 100), p: 0, float: 10000.0, double: 10000.0, difference: 0.0
input size (100, 100), p: 1, float: 11474.185546875, double: 11474.185433543797, difference: 0.00011333120346534997
input size (100, 100), p: 2, float: 143.50733947753906, double: 143.5073391487933, difference: 3.2874575595087663e-07
input size (100, 100), p: 3, float: 36.04550552368164, double: 36.045502752127405, difference: 2.7715542358919265e-06
input size (100, 100), p: 4, float: 18.796096801757812, double: 18.796098078653177, difference: 1.2768953645547754e-06
input size (100, 100), p: inf, float: 5.540258407592773, double: 5.5402586460113525, difference: 2.384185791015625e-07
input size (100, 100), p: -inf, float: 3.4868717193603516e-06, double: 3.4868717193603516e-06, difference: 0.0
```

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit a0c4a67c:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a0c4a67cd0dddb282169ae0652cd5e3764b7db06).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4140324) pytorch_linux_xenial_py2_7_9_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80741729) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80741729))
<details>
<summary>
<code>Jan 06 15:04:45 RuntimeError: test_torch failed!</code>
</summary>

```
Jan 06 15:04:45 Ran 1019 tests in 78.119s 
Jan 06 15:04:45  
Jan 06 15:04:45 FAILED (failures=1, skipped=46) 
Jan 06 15:04:45  
Jan 06 15:04:45 Generating XML reports... 
Jan 06 15:04:45 Traceback (most recent call last): 
Jan 06 15:04:45   File "test/run_test.py", line 456, in <module> 
Jan 06 15:04:45     main() 
Jan 06 15:04:45   File "test/run_test.py", line 449, in main 
Jan 06 15:04:45     raise RuntimeError(message) 
Jan 06 15:04:45 RuntimeError: test_torch failed! 
Jan 06 15:04:46 + cleanup 
Jan 06 15:04:46 + retcode=1 
Jan 06 15:04:46 + set +x 
Jan 06 15:04:46 =================== sccache compilation log =================== 
Jan 06 15:04:46 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Jan 06 15:04:46  
Jan 06 15:04:46 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 06 15:04:46 Compile requests                 51 
Jan 06 15:04:46 Compile requests executed        27 
Jan 06 15:04:46 Cache hits                        3 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4140321) pytorch_linux_xenial_py3_6_gcc5_4_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80741806) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80741806))
<details>
<summary>
<code>Jan 06 15:05:04 RuntimeError: test_torch failed!</code>
</summary>

```
Jan 06 15:05:04 Ran 1019 tests in 74.842s 
Jan 06 15:05:04  
Jan 06 15:05:04 FAILED (failures=1, skipped=36) 
Jan 06 15:05:04  
Jan 06 15:05:04 Generating XML reports... 
Jan 06 15:05:04 Traceback (most recent call last): 
Jan 06 15:05:04   File "test/run_test.py", line 456, in <module> 
Jan 06 15:05:04     main() 
Jan 06 15:05:04   File "test/run_test.py", line 449, in main 
Jan 06 15:05:04     raise RuntimeError(message) 
Jan 06 15:05:04 RuntimeError: test_torch failed! 
Jan 06 15:05:04 + cleanup 
Jan 06 15:05:04 + retcode=1 
Jan 06 15:05:04 + set +x 
Jan 06 15:05:04 =================== sccache compilation log =================== 
Jan 06 15:05:04 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Jan 06 15:05:04  
Jan 06 15:05:04 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 06 15:05:04 Compile requests                 48 
Jan 06 15:05:04 Compile requests executed        25 
Jan 06 15:05:04 Cache hits                        1 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by albanD** [comment]
You will need to rebase as there are conflicts with master.
Also the CI failure look relevant. Can you take a look?

**Comment by rivergold** [comment]
> You will need to rebase as there are conflicts with master.
> Also the CI failure look relevant. Can you take a look?

I have rebased and updated the code, but there is still a failure. Can give some guidances？Thanks a lot.

**Comment by albanD** [comment]
You can ignore the macos build failure.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -4906,7 +4906,7 @@
   use_c10_dispatcher: full
   variants: method, function
   dispatch:
-    CPU: legacy::cpu::_th_dist
+    CPU: dist
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: Is the implementation in TH used by anything else? If no, could you delete it please?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -4906,7 +4906,7 @@
...
     CUDA: legacy::cuda::_th_dist
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: Your implementation work both for cpu and cuda, is there a specific reason why you left the THC implementation for cuda?

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -967,4 +967,9 @@ std::tuple<Tensor, Tensor> cummin(const Tensor& self, Dimname dim) {
 std::tuple<Tensor&, Tensor&> cummin_out(Tensor& values, Tensor& indices, const Tensor& self, Dimname dim) {
   return at::cummin_out(values, indices, self, dimname_to_position(self, dim));
 }
+
+Tensor dist(const Tensor &self, const Tensor& other, Scalar p){
+  return at::norm(self - other, p);
```
Comment (albanD) [inline aten/src/ATen/native/ReduceOps.cpp]: Could you do a quick benchmark to make sure there are no performance regressions?
Also the implementation in TH was returning high precision result (double if the input is float. long if the input is int, etc). Does this implementation has the same behavior or does it have lower precision?

**Review by albanD — COMMENTED** [review]
Thanks for PR,
Just some concerns about avoiding any regression for speed and accuracy of the result.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (rivergold) [reply aten/src/ATen/native/native_functions.yaml]: It's my first time to make PR for PyTorch, and reference to [Refactor philosophy](https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide#refactor-philosophy) to do one thing a time. So I just change `dist` for CPU. After I finish the check of performance regression, I will update the code for cuda.

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml]: Sure, no worries. I am fine if the final PR only contains the CPU update. My question was really to know if it was just out of scope of the PR or something else that prevented you from doing it.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 3 above)
Comment (rivergold) [reply aten/src/ATen/native/ReduceOps.cpp]: Benchmark has been added. 
TH or native implementation only support float type, so there is no high precision result of int to long.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (rivergold) [reply aten/src/ATen/native/native_functions.yaml]: Has deleted TH dist cpu implementation

**Comment by albanD** [comment]
Looks good.
For precision I meant in terms of value. Because this is floating point numbers, the result is not exact. And for large Tensor, we don't want the result to be off by too much. In particular, the TH implementation for `float32`, the TH implementation accumulates in a `double` to reduce precision loss. I was wondering of `norm()` does that too or not.

**Comment by VitalyFedyunin** [comment]
Looks good to me. BTW you can put CPU and CUDA implementations together (at least it seems so).

**Comment by rivergold** [comment]
> Looks good.
> For precision I meant in terms of value. Because this is floating point numbers, the result is not exact. And for large Tensor, we don't want the result to be off by too much. In particular, the TH implementation for `float32`, the TH implementation accumulates in a `double` to reduce precision loss. I was wondering of `norm()` does that too or not.

Can you provide some references or examples as guidance for me to do a test? I can't find an example in test_torch.py. Thanks a lot!

**Comment by albanD** [comment]
I was thinking of running something like
```python
import torch

torch.manual_seed(123)
a = torch.rand(100, 100, 100)
b = torch.rand(100, 100, 100)

print(torch.dist(a, b))
print(torch.dist(a.double(), b.double()))
```
With the previous and new code. To see if the different is much bigger with the new code or not.

As @VitalyFedyunin mentioned, even though it returns an `accreal` which is higher precision, it still accumulates in regular precision `scalar_t sum;`. So you should see similar results :)

**Comment by rivergold** [comment]
@albanD  @VitalyFedyunin 
CPU and Cuda implementations are updated together.

**Comment by albanD** [comment]
Thanks. Could you rebase on top of master to make sure all test pass please?

Also did you had a chance to run the precision test above with this PR?

**Comment by rivergold** [comment]
@albanD I have finished rebasing and precision testing and the precision result is as what you mentioned. But here is still a build failed without any identified problem, please help me look at this problem.

**Comment by albanD** [comment]
Most likely a flaky test. I relaunched them and we'll see !

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**